File: challenge/series/series.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double geometric(int);
double series(double (*)(int), int, int);

int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: series MIN MAX\n");
        exit(EXIT_FAILURE);
    }

    int min = atoi(argv[1]);
    int max = atoi(argv[2]);
    double (*fn)(int) = geometric;

    double sum = series(fn, min, max);
    printf("%f\n", sum);
    return 0;
}

double geometric(int n) {
    return pow(0.25, n);
}

double series(double (*fn)(int), int min, int max) {
    double sum = 0;
    for (int i = min; i <= max; ++i) {
        sum += fn(i);
    }
    return sum;
}

File: challenge/series/Series.java
import java.util.stream.IntStream;
import java.util.function.IntToDoubleFunction;

public class Series {
    public static void main (String[] args) {
        if (args.length != 2) {
            System.out.println("Usage: Series MIN MAX");
            return;
        }
        int min = Integer.parseInt(args[0]);
        int max = Integer.parseInt(args[1]);
        
//        double sum = series(Series::geometric, min, max);
//        System.out.println(sum);
//
        double sum = series(n -> Math.pow(0.25, n), min, max);
        System.out.println(sum);
    }

    private static double geometric(int n) {
        return Math.pow(1.0 / 4.0, n);
    }
    
    private static Double series(IntToDoubleFunction fn, 
            int low, int high) {
        return IntStream.range(low, high).mapToDouble(fn).sum();
    }

//// generic approach, with 'import java.util.function.Function;'
//    private static Double series(Function<Integer, Double> fn, 
//            int low, int high) {
//        return IntStream.range(low, high)
//                        .mapToDouble(n -> fn.apply(n))
//                        .sum();
//    }

}
